updated SI-7331, SI-6843, SI-7731, parser entry point refactoring, assertThrows utility function

densh opened the pull request:
This is a rehash of #2808 that addresses feedback from @gkossakowski and @retronym

densh committed:
refactor parser entry points and extract a few methods out

This commit contains three logical changes:

1. Split `templateStatSeq` into two methods as we need more reliable
   parsing of template body alone for new `parseStats` entry point.

2. Add new parser entry point called `parseStats` which is aimed towards
   use in tools that require parsing of Scala code that can be written
   inside of a template.

   Such functionality is required for parsing lines in repl, parsing
   code through toolbox, parsing and running scala scripts and lastly
   for quasiquotes. All of them are refactored to use this very method
   in the next commits.

   A new method called `templateStatsCompat` is also added to make this
   commit pass the tests but it's a temporary hack that will be removed
   in next commit in favor of `parseStats`.

3. Extract out a few methods like `isCaseDefStart`, `expectedMsgTemplate`
   and `parseRule`. These are needed to override parser behaviour in
   updated quasiquotes parser (see next commits).

densh committed:
refactor repl to use new new parser entry point

This commit refactors repl to use `parseStats` entry point and
streamlines hacky error handling that was previously used to
encode errors that happen during parsing.

densh committed:
SI-7731 make CannotHaveAttrs more consistent

Previously setPos, pos_=, setType, tpe_= all behaved inconsistently
between each other even though they all represent similar
attributes that cannot be changed on CannotHaveAttrs trees.

In order to simplify handling of such trees in compiler code each
of these fields now supports assignment to its current default
value: NoType for tpe and NoPosition for pos. Such assignments
don't mutate underlying trees.

densh committed:
SI-7331 remove all the wrapping code from toolbox

Change toolbox parsing to use `parseStats` parser entry point
instead of current code-wrappign technique that makes positions
much less useful to end users.

There is also no need to create a compiler `Run` for parsing.

densh committed:
SI-6843 well-positioned syntax errors for quasiquotes

This is achieved in a following way:

1. Similarly to toolbox quasiquotes can go away with wrapping for
   parsing purpose after introduction of `parseStats` and `parseRule`
   entry points.

2. In case of syntax error quasiquote computes equivalent corresponding
   position in the source code with the help of `corrrespondingPosition`
   mapper which relies on position data collected into `posMap` during
   code generation.

densh committed:
add assertThrows testing utility function

This is much easier to use than built-in JUnit method-level checks.

adriaanm commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 22653839)
:cat: Roger! Rebuilding pr-scala for a06a771d, d7ad291a, 5439c4c9, b4598b4a, e1bef09d, 834e29ff. :rotating_light:

gkossakowski commented:
@densh: who is reviewer fro this PR?

densh commented:
@gkossakowski I guess that @retronym is the reviewer.

retronym commented on src/compiler/scala/tools/reflect/quasiquotes/Placeholders.scala:
```diff
@@ -17,18 +17,31 @@ trait Placeholders { self: Quasiquotes =>
   // Step 1: Transform Scala source with holes into vanilla Scala source
 
   lazy val holeMap = new HoleMap()
+  lazy val posMap = mutable.ListMap[Position, (Int, Int)]()
```
What is is the lifecycle of this map?

densh replied:
It gets initialized during code generation and lays dormant until syntax error happens. In that case it's used to determine equivalent position between quasiquote snippet and code that is fed to the quasiquote parser.

retronym replied:
So basically for the lifetime of a single call to one of the quasiquote macros. I was just checking that this wasn't mixed in to Global or something similar.

```
make(  QuasiquoteClass_api_apply) { case _                                          => _.expandQuasiquote },
    make(QuasiquoteClass_api_unapply) { case _                                          => _.expandQuasiquote }
```

densh replied:
Yes, exactly. It definitely doesn't get mixed into global.

retronym commented on test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala:
```diff
@@ -0,0 +1,70 @@
+    }
+    val attrlessTrees = List(CHA, EmptyTree, emptyValDef, pendingSuperCall)
+  }
+  def withCtx(body: CustomSymbolTable => Unit) = body(new CustomSymbolTable)
```
JUnit creates a new instance of the test for each `@Test` annotated method. (http://martinfowler.com/bliki/JunitNewInstance.html) So you can simple make `CustomSymbolTable` an object and lose the `withCtx` stuff. /cc @gkossakowski

densh replied:
Good to know, I'll update the tests now.

retronym commented on test/junit/scala/tools/testing/AssertUtil.scala:
```diff
@@ -0,0 +1,19 @@
+ *  that are ultimately based on junit.Assert primitives.
+ */
+object AssertUtil {
+  /** Check if exception T was thrawn during evaluation of f. If any
```
"thrown"

retronym replied:
"Check if exception T (or a subclass)"

densh committed:
addresses feedback regarding new junit tests

1. don't recreate symbol table manually and just let JUnit do it
   automatically behind the scenes
2. minor changes to assertThrow's description
3. add one more test case to check that exception's subclasses
   are catched too
4. refine rethrow test to ensure that the exception wasn't swallowed

gkossakowski commented:
I didn't realize JUnit is smart about class's life cycle. Thanks for the tip @retronym.

som-snytt commented:
https://github.com/junit-team/junit/wiki/Test-fixtures

I recall (years ago anyway), for N tests, your N instances are created simultaneously, whereas setUp means before test run.

retronym commented:
LGTM. Great work.
